pkg/sql/unsplit_range_test.go

```diff
@@ -0,0 +1,320 @@
+		if err != nil {
+			continue
+		}
+		if indexID == descpb.IndexID(foundIndexID) && !desc.GetStickyBit().IsEmpty() {
+			return true
+		}
+	}
+
+	return false
+}
+
+// splitLastRangeInSpan split the last range into two ranges,
+// and check the new range has sticky bit set.
+func splitLastRangeInSpan(
+	ctx context.Context, t *testing.T, kvDB *kv.DB, tableSpan roachpb.Span,
+) roachpb.Key {
+	metaStartKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.Key))
+	metaEndKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.EndKey))
+	ranges, err := kvDB.Scan(ctx, metaStartKey, metaEndKey, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lastRange := ranges[len(ranges)-1]
+	var lastRangeDesc roachpb.RangeDescriptor
+	if err := lastRange.ValueProto(&lastRangeDesc); err != nil {
+		t.Fatal(err)
+	}
+	splitKey := getMidKeyInSpan(t, kvDB, lastRangeDesc.StartKey, lastRangeDesc.EndKey)
+	splitKey, err = keys.EnsureSafeSplitKey(splitKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := kvDB.AdminSplit(ctx, splitKey, hlc.MaxTimestamp); err != nil {
+		t.Fatal(err)
+	}
+
+	return splitKey
+}
+
+// rangeIsManuallySplit check if there is a range starts with `startKey` and is
+// manually split
+func rangeIsManuallySplit(
+	ctx context.Context, t *testing.T, kvDB *kv.DB, tableSpan roachpb.Span, startKey roachpb.Key,
+) bool {
+	metaStartKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.Key))
+	metaEndKey := keys.RangeMetaKey(keys.MustAddr(tableSpan.EndKey))
+	ranges, err := kvDB.Scan(ctx, metaStartKey, metaEndKey, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.NotEmpty(t, ranges)
+	for _, r := range ranges {
+		var desc roachpb.RangeDescriptor
+		if err := r.ValueProto(&desc); err != nil {
+			t.Fatal(err)
+		}
+		if bytes.Equal(desc.StartKey.AsRawKey(), startKey) && !desc.GetStickyBit().IsEmpty() {
+			return true
+		}
+	}
+	return false
+}
+
+// Test that manually split ranges get unsplit when dropping a
+// table/database/index or truncating a table It verifies that the logic is
+// working on both the old (before version `UnsplitRangesInAsyncGCJobs`) and new
+// (from versiom `UnsplitRangesInAsyncGCJobs`) pathes.
+// TODO(Chengxiong): remove test for test cases with binary version
+// `clusterversion.UnsplitRangesInAsyncGCJobs - 1` and update this comment in
+// 22.2
+func TestUnsplitRanges(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	const numRows = 2*row.TableTruncateChunkSize + 1
+	const numKeys = 3 * numRows
+	const tableName string = "test1"
+
+	testCases := []struct {
+		name                   string
+		query                  string
+		binaryVersion          clusterversion.Key
+		allKeyCntAfterGC       int  // Key count with table prefix.
+		indexKeyCntAfterGC     int  // Key count with index prefix.
+		hasSplitOnTableAfterGC bool // Any split on the table in general?
+		hasSplitOnKeyAfterGC   bool // Any split on the range start with splitKey (where we manually split in the test)?
```

Comment:
I think you could use commentary to explain to a reader better how this test works. Try to come to this test with the idea that you know nothing about how it works or what it does and try to make the commentary useful enough to make it not too painful to figure out.